### PR TITLE
IDVA51337-- [Registration] Sole Trader Only IDV - Feature Flag

### PR DESCRIPTION
**Pull request by rdas-ch**

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/app.ts [hunk 1]
```diff
@@ -55,6 +56,7 @@ app.use(express.static(path.join(__dirname, "/../assets/public")));
 app.use(cookieParser());
 app.use(`^(?!(${BASE_URL}${HEALTHCHECK}|${BASE_URL}$|${BASE_URL}${ACCESSIBILITY_STATEMENT}))*`, sessionMiddleware);
 app.use(`^(?!(${BASE_URL}${HEALTHCHECK}|${BASE_URL}$|${BASE_URL}${ACCESSIBILITY_STATEMENT}))*`, authenticationMiddleware);
```
Comment (sampankumar) [inline src/app.ts:58]: I think you will have to ignore sole trader urls to do this

File: src/middleware/authentication_middleware_sole_trader.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+import { AuthOptions, acspProfileCreateAuthMiddleware } from "@companieshouse/web-security-node";
+
+import { CHS_URL } from "../utils/properties";
+import { BASE_URL, CHECK_SAVED_APPLICATION } from "../types/pageURL";
+
+export const authenticationMiddlewareForSoleTrader = (req: Request, res: Response, next: NextFunction) => {
+
+    const authMiddlewareConfig: AuthOptions = {
+        chsWebUrl: CHS_URL,
+        returnUrl: BASE_URL + CHECK_SAVED_APPLICATION
```
Comment (sampankumar) [inline src/middleware/authentication_middleware_sole_trader.ts:11]: Return url would be different than check saved application if only sole trader verification is enabled.

File: src/middleware/authentication_middleware.ts [hunk 3]
```diff
@@ -11,5 +11,7 @@ export const authenticationMiddleware = (req: Request, res: Response, next: Next
         returnUrl: BASE_URL + CHECK_SAVED_APPLICATION
     };
 
-    return acspProfileCreateAuthMiddleware(authMiddlewareConfig)(req, res, next);
+    return FEATURE_FLAG_DISABLE_MANDATORY_VERIFICATION === "true"
```
Comment (sampankumar) [inline src/middleware/authentication_middleware.ts:14]: May be use isActiveFeature function here.

File: src/utils/properties.ts [hunk 4]
```diff
@@ -49,3 +49,5 @@ export const PIWIK_START_GOAL_ID = getEnvironmentValue("PIWIK_START_GOAL_ID", "4
 export const FEATURE_FLAG_DISABLE_LIMITED_JOURNEY = getEnvironmentValue("FEATURE_FLAG_DISABLE_LIMITED_JOURNEY", "false");
 
 export const FEATURE_FLAG_DISABLE_PARTNERSHIP_JOURNEY = getEnvironmentValue("FEATURE_FLAG_DISABLE_PARTNERSHIP_JOURNEY", "false");
+
+export const FEATURE_FLAG_DISABLE_MANDATORY_VERIFICATION = getEnvironmentValue("FEATURE_FLAG_DISABLE_MANDATORY_VERIFICATION", "true");
```
Comment (swalke-ch) [inline src/utils/properties.ts:53]: I think this need to be renamed to FEATURE_FLAG_VERIFY_SOLE_TRADER_ONLY

File: src/app.ts (hunk 1 above)
Comment (sampankumar) [reply src/app.ts:58]: I still believe you need to ignore sole trader urls here

File: src/utils/properties.ts [hunk 5]
```diff
@@ -49,3 +49,5 @@ export const PIWIK_START_GOAL_ID = getEnvironmentValue("PIWIK_START_GOAL_ID", "4
 export const FEATURE_FLAG_DISABLE_LIMITED_JOURNEY = getEnvironmentValue("FEATURE_FLAG_DISABLE_LIMITED_JOURNEY", "false");
 
 export const FEATURE_FLAG_DISABLE_PARTNERSHIP_JOURNEY = getEnvironmentValue("FEATURE_FLAG_DISABLE_PARTNERSHIP_JOURNEY", "false");
+
+export const FEATURE_FLAG_VERIFY_SOLE_TRADER_ONLY = getEnvironmentValue("FEATURE_FLAG_VERIFY_SOLE_TRADER_ONLY", "true");
```
Comment (swalke-ch) [inline src/utils/properties.ts:53]: The default value should be false? We want to verify everyone by default.

File: src/middleware/authentication_middleware.ts [hunk 6]
```diff
@@ -11,5 +11,7 @@ export const authenticationMiddleware = (req: Request, res: Response, next: Next
         returnUrl: BASE_URL + CHECK_SAVED_APPLICATION
     };
 
-    return acspProfileCreateAuthMiddleware(authMiddlewareConfig)(req, res, next);
+    return FEATURE_FLAG_VERIFY_SOLE_TRADER_ONLY === "true"
```
Comment (swalke-ch) [inline src/middleware/authentication_middleware.ts:14]: functionality need to be reversed, as mentioned in the other comment, if the flag is true then verify sole trader only and not the other way round

File: src/middleware/authentication_middleware.ts (hunk 3 above)
Comment (rdas-ch) [reply src/middleware/authentication_middleware.ts]: done. Thanks

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/utils/properties.ts (hunk 5 above)
Comment (rdas-ch) [reply src/utils/properties.ts]: Yes, made it false as default

File: src/middleware/authentication_middleware.ts [hunk 7]
```diff
@@ -11,5 +11,7 @@ export const authenticationMiddleware = (req: Request, res: Response, next: Next
         returnUrl: BASE_URL + CHECK_SAVED_APPLICATION
     };
 
-    return acspProfileCreateAuthMiddleware(authMiddlewareConfig)(req, res, next);
+    return isActiveFeature(FEATURE_FLAG_VERIFY_SOLE_TRADER_ONLY) === true
```
Comment (swalke-ch) [inline src/middleware/authentication_middleware.ts:14]: If you are calling `isActiveFeature()` then you don't need check `=== true`

File: src/app.ts [hunk 8]
```diff
@@ -54,7 +55,8 @@ app.use(express.static(path.join(__dirname, "/../assets/public")));
 // Apply middleware
 app.use(cookieParser());
 app.use(`^(?!(${BASE_URL}${HEALTHCHECK}|${BASE_URL}$|${BASE_URL}${ACCESSIBILITY_STATEMENT}))*`, sessionMiddleware);
-app.use(`^(?!(${BASE_URL}${HEALTHCHECK}|${BASE_URL}$|${BASE_URL}${ACCESSIBILITY_STATEMENT}))*`, authenticationMiddleware);
+app.use(`^(?!(${BASE_URL}${HEALTHCHECK}|${BASE_URL}$|${BASE_URL}${ACCESSIBILITY_STATEMENT})|(${BASE_URL}${SOLE_TRADER_WHAT_IS_YOUR_ROLE}))*`, authenticationMiddleware);
+app.use(`^(${BASE_URL}${SOLE_TRADER_WHAT_IS_YOUR_ROLE})*`, authenticationMiddlewareForSoleTrader);
```
Comment (swalke-ch) [inline src/app.ts:59]: this middleware needs to apply to all the sole trader urls

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/middleware/authentication_middleware.ts (hunk 7 above)
Comment (rdas-ch) [reply src/middleware/authentication_middleware.ts]: done, thanks

File: src/app.ts (hunk 8 above)
Comment (rdas-ch) [reply src/app.ts]: Done, thank you

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by ttingle-ch — APPROVED** [review]
LGTM

**Review by sampankumar — APPROVED** [review]

**Review by swalke-ch — APPROVED** [review]

**Pull request merged**